### DISCUSSION TITLE
What does "config needs migrating" mean?

### DISCUSSION THREAD

**Discussion by amimas** [renovatebot/renovate#10179] [Request Help]
When I run the renovate bot, I see the following warning about `config needs migrating`. What does it mean?

When I look at both `originalConfig` and `migratedConfig` from the output, I can't see any difference. Not sure what's causing this and if I should be concerned. I've redacted some lines with `****` in the following output but in my console, I don't see any difference in the values being printed.

```
 WARN: Config needs migrating
       "originalConfig": {
         "platform": "gitlab",
         "endpoint": "****",
         "token": "***********",
         "autodiscover": true,
         "binarySource": "global",
         "gitAuthor": "*****",
         "force": {"automerge": false, "enabled": true},
         "logFile": "renovate.log",
         "logFileLevel": "debug",
         "onboarding": false,
         "optimizeForDisabled": true,
         "persistRepoData": false,
         "prCommitsPerRunLimit": 0,
         "printConfig": true,
         "requireConfig": false,
         "extends": ["config:base"],
         "assigneesFromCodeOwners": true,
         "automerge": false,
         "commitBodyTable": true,
         "enabled": true,
         "hostRules": [
           {"matchHost": "docker.io", "authType": "Bearer"},
           {"matchHost": "quay.io", "authType": "Bearer"}
         ],
         "prFooter": "This MR has been generated by [Renovate Bot](https://****/renovatebot/)",
         "schedule": []
       },
       "migratedConfig": {
         "platform": "gitlab",
         "endpoint": "******",
         "token": "***********",
         "autodiscover": true,
         "binarySource": "global",
         "gitAuthor": "*****",
         "force": {"automerge": false, "enabled": true},
         "logFile": "renovate.log",
         "logFileLevel": "debug",
         "onboarding": false,
         "optimizeForDisabled": true,
         "persistRepoData": false,
         "prCommitsPerRunLimit": 0,
         "printConfig": true,
         "requireConfig": false,
         "extends": ["config:base"],
         "assigneesFromCodeOwners": true,
         "automerge": false,
         "commitBodyTable": true,
         "enabled": true,
         "hostRules": [
           {"matchHost": "docker.io", "authType": "Bearer"},
           {"matchHost": "quay.io", "authType": "Bearer"}
         ],
         "prFooter": "This MR has been generated by [Renovate Bot](https://*****/renovatebot/)",
         "schedule": []
       }
```

**Comment by rarkins** [comment]
Have you inserted any or all of the `***` sanitization above, or was that all done by Renovate? I don't see any other differences:

![image](https://user-images.githubusercontent.com/6311784/119719036-97f19400-be68-11eb-9c63-2a283da604e6.png)

**Reply by amimas** [reply]
I *think* I've figured out when this happens after a observing several different executions.

The `config.js` for me uses environment variable for sensitive info (i.e. secrets, tokens, credentials.). For example: `token: process.env.RENOVATE_TOKEN`. I have several other configs (i.e. secrets) that are not saved in the `config.js` as plain text. I noticed this migration warning show up when one or more info is expected by `config.js` but is not set in the expected environment variable.

Not sure what happens when the tool performs "migration". It's good that the tool is able to detect the absence of required info. I wonder if the message can be further clarified in this scenario, if what I observed is accurate.

**Reply by HenrikPoulsen** [reply]
I have the same issue for my self-hosted renovate.
The originalConfig and migratedConfig are identical according to all known diffing tools to mankind. I dont have any passwords or anything in there so the config is truly identical.
I have some sanity checks in CI whenever I update renovate to make sure we don't get any new warnings printed when I run it in dry run, but this particular one leads to some hacks so that warning doesn't count.
Would be nice to know what is causing it

**Reply by rarkins** [reply]
Can you reproduce it in a public repo? Alternatively, duplicate it in a private one and then reduce the config until it no longer reproduces?

**Reply by HenrikPoulsen** [reply]
Sorry for not replying here. I was unable to reproduce it locally. But I suspect it likely has something to do with the access tokens that get generated at runtime and gets used by renovate through environment variables that are set in the config.js. Haven't veritied this though. Just a guess at this point.
Those values get obfuscated when renovate prints the config, which would explain why the output is identical when checking it manually.
No idea if it's reasonable for that warning to use the obfuscated values during comparison, which would basically ignore them.
In my case I worked around it by downgrading the warning to an INFO print. Which obviously isn't great but it allows me to keep warnings as errors in my setup while it still prints those messages in most of the runs
